backend/cmm_helpers.ml

```diff
@@ -2248,22 +2248,37 @@ let arraylength kind arg dbg =
    Check if this can be improved (e.g., bswap). *)
 
 let bbswap bi arg dbg =
-  let prim, tyarg = match (bi : Primitive.boxed_integer) with
-    | Pnativeint -> "nativeint", XInt
-    | Pint32 -> "int32", XInt32
-    | Pint64 -> "int64", XInt64
+  let bitwidth : Cmm.bswap_bitwidth =
+    match (bi : Primitive.boxed_integer) with
+    | Pnativeint -> if size_int = 4 then Thirtytwo else Sixtyfour
+    | Pint32 -> Thirtytwo
+    | Pint64 -> Sixtyfour
   in
-  Cop(Cextcall { func = Printf.sprintf "caml_%s_direct_bswap" prim;
-                 builtin = false;
-                 returns = true;
-                 effects = Arbitrary_effects;
-                 coeffects = Has_coeffects;
-                 ty = typ_int; alloc = false; ty_args = [tyarg]; },
-      [arg],
-      dbg)
+  let op = Cbswap { bitwidth } in
+  if (bi = Primitive.Pint64 && size_int = 4) ||
+     not (Proc.operation_supported op) then
+    let prim, tyarg = match (bi : Primitive.boxed_integer) with
+      | Pnativeint -> "nativeint", XInt
+      | Pint32 -> "int32", XInt32
+      | Pint64 -> "int64", XInt64
+    in
+    Cop(Cextcall { func = Printf.sprintf "caml_%s_direct_bswap" prim;
+                   builtin = false;
+                   returns = true;
+                   effects = Arbitrary_effects;
+                   coeffects = Has_coeffects;
+                   ty = typ_int; alloc = false; ty_args = [tyarg]; },
```

Comment:
Shouldn't the `ty` field be set to `tyarg`?

Reply:
`ty` is of type `machtype` but `tyarg` is of type `exttype` which is confusingly completely separate.

Reply:
Sorry for the mistake, but my question was more along
the lines of: why is the argument type variable but the
return type fixed? I was expecting everything to be
`t -> t` with `t` any boxed integer type.

Reply:
Or is it because we are assured the result is always unboxed?

Reply:
The result is always unboxed, so the `machtype` `typ_int` is correct. The arguments are always unboxed too.  This new `ty_args` is not relevant for amd64. See https://github.com/ocaml/ocaml/pull/9752 bullet 3 in the proposed approach.

Reply:
Noted, thanks.